src/lib.rs

```diff
@@ -640,33 +642,198 @@ fn rdo_mode_decision(fi: &FrameInvariants, fs: &mut FrameState,
+                let r = ((cw.w.tell_frac() - tell) as f64)/8.0;
+                let sub_rd = (d as f64) + lambda * r - rd;
+                rd += sub_rd;
+
+                child_modes.push(RDOPartitionOutput { bo: offset.clone(), pred_mode: mode, rd_cost: sub_rd as u64 });*/
+
+
+                let offset = BlockOffset{x: bo.x, y: bo.y + hbs as usize};
+                let mode = rdo_mode_decision(fi, fs, cw, subsize, &offset).part_modes[0].pred_mode;
+                cw.bc.set_mode(&offset, subsize, mode);
+                encode_block(fi, fs, cw, mode, subsize, &offset);
+
+                /*let po = offset.plane_offset(&fs.input.planes[0].cfg);
+                let d = sse_wxh(&fs.input.planes[0].slice(&po), &fs.rec.planes[0].slice(&po),
+                                subw as usize, subh as usize);
+                let r = ((cw.w.tell_frac() - tell) as f64)/8.0;
+                let sub_rd = (d as f64) + lambda * r - rd;
+                rd += sub_rd;
+
+                child_modes.push(RDOPartitionOutput { bo: offset.clone(), pred_mode: mode, rd_cost: sub_rd as u64 });*/
+
+
+                let offset = BlockOffset{x: bo.x + hbs as usize, y: bo.y + hbs as usize};
+                let mode = rdo_mode_decision(fi, fs, cw, subsize, &offset).part_modes[0].pred_mode;
+                cw.bc.set_mode(&offset, subsize, mode);
+                encode_block(fi, fs, cw, mode, subsize, &offset);
+
+                /*let po = offset.plane_offset(&fs.input.planes[0].cfg);
+                let d = sse_wxh(&fs.input.planes[0].slice(&po), &fs.rec.planes[0].slice(&po),
+                                subw as usize, subh as usize);
+                let r = ((cw.w.tell_frac() - tell) as f64)/8.0;
+                let sub_rd = (d as f64) + lambda * r - rd;
+                rd += sub_rd;
+
+                child_modes.push(RDOPartitionOutput { bo: offset.clone(), pred_mode: mode, rd_cost: sub_rd as u64 });*/
+            },
+            _ => { assert!(false); },
+        }
+
+        if rd == max_rd {
+            let po = bo.plane_offset(&fs.input.planes[0].cfg);
+            let d = sse_wxh(&fs.input.planes[0].slice(&po), &fs.rec.planes[0].slice(&po),
+                            w as usize, h as usize);
+            let r = ((cw.w.tell_frac() - tell) as f64)/8.0;
+
+            rd = (d as f64) + lambda * r;
+        }
+        
+        if rd < best_rd {
+            best_rd = rd;
+            best_partition = partition;
+            best_pred_modes = child_modes.clone();
+        }
+
+        cw.rollback(checkpoint.clone());
+    }
+
+    assert!(best_rd as i64 >= 0);
+    
     let rdo_output = RDOOutput { rd_cost: best_rd as u64,
-                                pred_mode: best_mode};
+                                part_type: best_partition,
+                                part_modes: best_pred_modes };
+
     rdo_output
 }
 
 fn encode_partition(fi: &FrameInvariants, fs: &mut FrameState, cw: &mut ContextWriter,
```

Comment:
The original logic remains in `encode_partition`. I believe it may be better to handle the case of fixed partition sizes through context parameters or arguments. The block at lines 827-834 controls when a decision is allowed to be made.

Reply:
Agree, for the option of fixed size partition, we can do it later in the separate commit.
Probably by putting, if fi.used_fixed_size_partition block in line 770, where if true then skip rdo_partition_decision() and set partition = PARTITION_SPLIT.